### PR TITLE
Replicated

### PR DESCRIPTION
**Pull request by erykciepiela**
`Replicated a b` which is lattice product as defined in "Introduction to Lattices and Order" by Davey and Priestley it's a replicated lattice according to another lattice with all corresponding elements of first lattice replicas are ordered according to second lattice.

Review request: I'm not sure about lines I marked with TODO, any help very welcome.

### CODE REVIEW

File: src/Algebra/Lattice/Replicated.hs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+-- | Replicated lattice according to another lattice. 
+-- | All corresponding elements of first lattice replicas are ordered according to second lattice.
+data Replicated a b = Replicated a b
+    deriving ( Eq, Ord, Show, Read, Data, Typeable, Generic, Functor, Foldable, Traversable
+            , Generic1
+            )
+
+instance (NFData a, NFData b) => NFData (Replicated a b) where
+    rnf (Replicated a b) = rnf (rnf a, rnf b) -- TODO
```
Comment (phadej) [inline src/Algebra/Lattice/Replicated.hs]: See e.g. https://github.com/phadej/lattices/blob/ae6c37705caf40e1536b57c6a144d252f5f519b0/src/Algebra/Lattice/Lexicographic.hs#L73

```haskell
rnf a `seq` rnf b
```

File: src/Algebra/Lattice/Replicated.hs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    (Replicated xa xb) /\ (Replicated ya yb) = Replicated (xa /\ ya) (xb /\ yb)
+
+instance (BoundedJoinSemiLattice a, BoundedJoinSemiLattice b) => BoundedJoinSemiLattice (Replicated a b) where
+    bottom = Replicated bottom bottom
+
+instance (BoundedMeetSemiLattice a, BoundedMeetSemiLattice b) => BoundedMeetSemiLattice (Replicated a b) where
+    top = Replicated top top
+
+instance (Universe a, Universe b) => Universe (Replicated a b) where
+    universe = [Replicated a b | a <- universe, b <- universe]
```
Comment (phadej) [inline src/Algebra/Lattice/Replicated.hs]: use interleaving combinators from `universe-base`, check https://hackage.haskell.org/package/universe-base-1.1.1/docs/Data-Universe-Helpers.html

File: src/Algebra/Lattice/Replicated.hs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+instance (Finite a, Finite b) => Finite (Replicated a b) where
+    -- ?
```
Comment (phadej) [inline src/Algebra/Lattice/Replicated.hs]: ... and here one could use non-interleaving list comprehension

File: src/Algebra/Lattice/Replicated.hs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+instance (QC.Arbitrary a, QC.Arbitrary b) => QC.Arbitrary (Replicated a b) where
+    arbitrary = Replicated <$> QC.arbitrary <*> QC.arbitrary
+    shrink (Replicated a b) = Replicated <$> QC.shrink a <*> QC.shrink b
```
Comment (phadej) [inline src/Algebra/Lattice/Replicated.hs]: this is wrong way to shrink pair-like structures.

See https://github.com/phadej/lattices/blob/ae6c37705caf40e1536b57c6a144d252f5f519b0/src/Algebra/Lattice/Lexicographic.hs#L134

**Comment by erykciepiela** [comment]
In my copy of the book it's page 42, definition 2.15 Products.

File: src/Algebra/Lattice/Replicated.hs (hunk 1 above)
Comment (erykciepiela) [reply src/Algebra/Lattice/Replicated.hs]: Thanks for the hint, corrected.

File: src/Algebra/Lattice/Replicated.hs (hunk 2 above)
Comment (erykciepiela) [reply src/Algebra/Lattice/Replicated.hs]: Used `+*+` for cartesian product of universes.

File: src/Algebra/Lattice/Replicated.hs (hunk 3 above)
Comment (erykciepiela) [reply src/Algebra/Lattice/Replicated.hs]: Done. Also added cardinality as product of cardinalities.

File: src/Algebra/Lattice/Replicated.hs (hunk 4 above)
Comment (erykciepiela) [reply src/Algebra/Lattice/Replicated.hs]: Yep, you're right, corrected.

**Comment by erykciepiela** [comment]
All changes applied as requested. Is there a chance to get this merged? Thank you.

**Comment by phadej** [comment]
Sorry for forgetting about this. This instance is exactly the instance we have for `(a, b)`, so we don't need a newtype.

**Pull request closed without merging**